Mathlib/Topology/MetricSpace/MetricSeparated.lean

```diff
@@ -1,38 +1,74 @@
 /-
-Copyright (c) 2021 Yury Kudryashov. All rights reserved.
+Copyright (c) 2021 Yury Kudryashov, Yaël Dillies. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Yury Kudryashov
+Authors: Yury Kudryashov, Yaël Dillies
 -/
 import Mathlib.Topology.EMetricSpace.Defs
 
 /-!
-# Metric separated pairs of sets
+# Metric separation
 
-In this file we define the predicate `Metric.AreSeparated`. We say that two sets in an (extended)
-metric space are *metric separated* if the (extended) distance between `x ∈ s` and `y ∈ t` is
-bounded from below by a positive constant.
+This file defines a few notions of separations of sets in a metric space.
 
-This notion is useful, e.g., to define metric outer measures.
--/
+## Main declarations
 
+* `Metric.IsSeparated`: A set `s` is `ε`-separated if its elements are pairwise at distance at least
+  `ε` from each other.
+* `Metric.AreSeparated`: Two sets `s` and `t` are separated if the distance between `x ∈ s` and
+  `y ∈ t` is bounded from below by a positive constant.
+-/
 
 open EMetric Set
+open scoped ENNReal
 
 noncomputable section
 
 namespace Metric
+variable {X : Type*} [PseudoEMetricSpace X] {s t : Set X} {ε δ : ℝ≥0∞} {x : X}
+
+/-!
+### Metric-separated sets
+
+In this section we define the predicate `Metric.IsSeparated` for `ε`-separated sets.
+-/
+
+/-- A set `s` is `ε`-separated if its elements are pairwise at distance at least `ε` from each
+other. -/
+def IsSeparated (ε : ℝ≥0∞) (s : Set X) : Prop := s.Pairwise (ε < edist · ·)
+
+protected lemma IsSeparated.empty : IsSeparated ε (∅ : Set X) := pairwise_empty _
+protected lemma IsSeparated.singleton : IsSeparated ε {x} := pairwise_singleton ..
+
+@[simp] lemma IsSeparated.of_subsingleton (hs : s.Subsingleton) : IsSeparated ε s := hs.pairwise _
+
+alias _root_.Set.Subsingleton.isSeparated := IsSeparated.of_subsingleton
+
+nonrec lemma IsSeparated.anti (hεδ : ε ≤ δ) (hs : IsSeparated δ s) : IsSeparated ε s :=
+  hs.mono' fun _ _ ↦ hεδ.trans_lt
+
+lemma IsSeparated.subset (hst : s ⊆ t) (hs : IsSeparated ε t) : IsSeparated ε s := hs.mono hst
+
+protected lemma IsSeparated.insert (hs : IsSeparated ε s) (h : ∀ y ∈ s, x ≠ y → ε < edist x y) :
```

Comment:
Perhaps an `insert_of_not_mem` which should be easy from `Pairwise.insert_of_symmetric_of_not_mem`

Reply:
Actually, it strikes me that `Pairwise.insert_of_symmetric_of_not_mem` is strictly weaker than `Pairwise.insert_of_symmetric` and that the `a ∉ s` is unnecessary. Maybe we should just delete it?

Reply:
It's weaker, but I think it's worth having around as a convenience lemma. Specifically see [`Set.pairwise_insert_of_symmetric_of_not_mem`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Pairwise/Basic.html#Set.pairwise_insert_of_symmetric_of_not_mem), which I think should stay (but has a bad name)

Reply:
Yes so the iff version is worth having, but the mpr direction is silly and we don't want it, right?

Reply:
Oh, I see what you mean. Yeah, agreed. (You might want to have the iff version for IsSeparated though!)

Reply:
Done

Reply:
Heh, a second after I sent my comment they appeared :)

Reply:
Live review 😄 

I'm deprecating `Pairwise.insert_of_symmetric_of_not_mem` in #23112